src/core/compiler.jl

```diff
@@ -339,106 +322,60 @@ Builds the output expression.
-    inner_function_name = main_body_names[:inner_function]
-    defaults_name = main_body_names[:defaults]
+    ctx = main_body_names[:ctx]
+    vi = main_body_names[:vi]
+    model = main_body_names[:model]
+    sampler = main_body_names[:sampler]
+    inner_function = main_body_names[:inner_function]
 
     # Arguments with default values
     args = model_info[:args]
     # Argument symbols without default values
     arg_syms = model_info[:arg_syms]
+    # Arguments namedtuple
+    args_nt = model_info[:args_nt]
     # Default values of the arguments
     whereparams = model_info[:whereparams]
-    tent_arg_defaults_nt = model_info[:tent_arg_defaults_nt]
     # Model generator name
-    outer_function_name = model_info[:name]
-    # Tentative list of parameter variables
-    tent_pvars_list = model_info[:tent_pvars_list]
-    # Tentative list of data variables
-    tent_dvars_list = model_info[:tent_dvars_list]
+    outer_function = model_info[:name]
     # Main body of the model
     main_body = model_info[:main_body]
 
-    if length(tent_dvars_list) == 0
-        tent_dvars_nt = :(NamedTuple())
-    else
-        nt_type = Expr(:curly, :NamedTuple, 
-            Expr(:tuple, QuoteNode.(tent_dvars_list)...), 
-            Expr(:curly, :Tuple, [:(Turing.Core.get_type($x)) for x in tent_dvars_list]...)
-        )
-        tent_dvars_nt = Expr(:call, :(Turing.namedtuple), nt_type, Expr(:tuple, tent_dvars_list...))
-    end
-    #= Does the following for each of the tentative dvars
-        local x
-        if isdefined(model.data, :x)
-            x = model.data.x
-        else
-            x = default_x
-        end
-    =#
     unwrap_data_expr = Expr(:block)
-    for var in tent_dvars_list
+    for var in arg_syms
+        temp_var = gensym(:temp_var)
+        varT = gensym(:varT)
         push!(unwrap_data_expr.args, quote
             local $var
-            if isdefined($model_name.data, $(QuoteNode(var)))
-                # The `Type{T}` arguments will always show up in the `model.data` named tuple.
-                # This is because we are not replacing the default value of these arguments with `nothing` but keeping the same value defined by by the user, e.g. `Float64`.
-                # So if the value is indeed correct, i.e. a type then it should just work
-                # If the value is not a type, i.e. `::Type{T} = 1` and the user doesn't pass it something for this argument, then it will give an error when constructing the model, which is a correct Julia error.
-                # This means that we don't need an expression for the default value of the `Type{T}` arguments in `model.defaults`.
-                if $model_name.data.$var isa Type && ($model_name.data.$var <: AbstractFloat || $model_name.data.$var <: AbstractArray)
-                    $var = Turing.Core.get_matching_type($sampler_name, $vi_name, $model_name.data.$var)
+            if isdefined($model.args, $(QuoteNode(var)))
```

Comment:
Do we still need `isdefined` if we are going to thorw an error if something is not passed in?

Reply:
Agreed. The `isdefined` is unnecessary as it is implied.